### ISSUE TITLE
embed an "audio player shortcode" at the top of a post

### ISSUE THREAD

**Issue by aschweigert** [publicmediaplatform/pmp-wordpress#43]
If a PMP story you're pulling into a Post happens to have "audio" attached to it, grab the URL's to each piece of attached audio and embed them as shortcodes at the top of the Post body.

docs for WP audio shortcode: http://codex.wordpress.org/Audio_Shortcode

**Comment by aschweigert** [comment]
This is the one bit of phase 2 we want to include in the scope for phase 4

**Comment by cavis** [comment]
So just to be clear on what to do here...

1. Check if the `story` has an attached audio file
 * in the phpsdk, use `$story->items('audio')`
2. If so, check for presence an enclosure link
 * something like `$audio->links('enclosure')`
 * if present, use the first one's `href` as the media file, and embed the shortcode
3. As a backup, check for a contentencoded attribute
 * this could be some sort of iframe representation of the media
 * something like `$audio->attributes->contentencoded`
 * embed as html at top of the post
 * note: you probably won't find any actual examples of this sort of audio in the PMP today

**Comment by rnagle** [comment]
Numbers 1 and 2 have been addressed in f7d1468c6bb9c74ef1d3fedbd82782ae8e5a5faf.

**Comment by cavis** [comment]
Confirmed for NPR and PRI stories-with-audio.

Cannot check APM until they address the non-binary-enclosures issue.

BUG: doesn't seem to work with PRX content.  Tried to pull [this story](https://support.pmp.io/proxy/sandbox/docs/38f33881-461c-4eae-a1af-15b1577ff301) and got no audio.

**Comment by rnagle** [comment]
It looks like the plugin fails to embed the shortcode for that story because the href does not point to an actual media file. Instead, it redirects to one. See: https://github.com/publicmediaplatform/pmp-wordpress/blob/master/inc/ajax.php#L263-L269

Should we drop this check and depend on href field eventually resolving to a suitable media file?

**Comment by cavis** [comment]
Hmm... may have to.  I know that PRX recently put in [an unrelated fix](https://github.com/PRX/pmp-importer.prx.org/pull/10) that will remove that "count".  But seems like it's only temporary, at best.

Would it be entirely crazy to do a HEAD request on the enclosure href, and attempt to interpret the `Content-Type` header?  Although it may be less simple in this case - that count.prx.org gives me a "301 Moved Permanently" to cms.prx.org, which then gives me a "302 Found" to s3.amazonaws.com.  And then S3 gives me a "403 Forbidden" on the HEAD request.  (Seems it only allows GET).

Hmmphh.

**Comment by cavis** [comment]
How about just trusting the `enclosure.type` field?  You can detect `audio/m3u` in NPR content there, and handle those separately.

**Comment by rnagle** [comment]
Excellent. Yes, I'll use the `enclosure.type` field for this instead. Thanks for the suggestion!

**Comment by rnagle** [comment]
So it looks like the WP media player doesn't work without a direct link to the media file. There goes that idea.

/shrug

**Comment by cavis** [comment]
Hmm... may need to loop @kookster in on this one.

**Comment by kookster** [comment]
We need to do a redirect for metrics, so I don't so how we will be getting away from that.
I am very surprised other publishers are not also using redirects in the same way (in talking to them, I thought they were).

**Comment by cavis** [comment]
Not suggesting you take them out - I think it's perfectly valid to use redirects.  And any dereferencing on the client side (the way we're currently dereferencing m3u's) would be a bad idea, in this case.

I'm more wondering how we might make the default WP player handle these redirects.  /me does some more reading.

**Comment by rnagle** [comment]
@kookster Just poking around and noticed something about the href for the audio file @cavis mentioned in his example above. It looks like, once the redirect url for the audio file finally resolves, S3 responds with a 403. See: https://gist.github.com/rnagle/a41b2caae15c72c905c7

This is *probably* unrelated to the issue of getting the WP media player to work with the audio file. Maybe.

Any thoughts on why I'm seeing the 403?

**Comment by kookster** [comment]
Not sure what is wrong with curl, but chrome downloads the mp3 from the enclosure url just fine:

https://cms.prx.org/pub/572dbb04874a32676784039e7afc71eb/0/web/audio_file/829012/broadcast/Moth1501Promo.mp3

**Comment by kookster** [comment]
Ah - the problem is the `-I` in your curl request - you are making a HEAD request.
Remove that, and the curl command works fine.
```
curl -L https://count.prx.org/redirect\?action\=request\&action_value\=%7B%22audioFileId%22%3A829012%2C%22pieceId%22%3A141178%7D\&location\=https%3A%2F%2Fcms.prx.org%2Fpub%2F572dbb04874a32676784039e7afc71eb%2F0%2Fweb%2Faudio_file%2F829012%2Fbroadcast%2FMoth1501Promo.mp3\&prx_story_id\=141178\&referrer\=https%3A%2F%2Fapi-sandbox.pmp.io%2Fdocs%2Fca7b2d27-f1e0-49ef-b831-b90a546cbfb5

```

**Comment by kookster** [comment]
Made a quick test page:
https://kookstertest.wordpress.com/2015/04/29/testing-audio-shortcodes/

Redirects work, as this redirects to s3:
https://cms.prx.org/pub/572dbb04874a32676784039e7afc71eb/0/web/audio_file/829012/broadcast/Moth1501Promo.mp3

and that is what urls will look like when I republish the audio, so I'm not so worried about it.

The count url (which is going away) doesn't work: "https://count.prx.org/redirect\?action\=request\&action_value\=%7B%22audioFileId%22%3A829012%2C%22pieceId%22%3A141178%7D\&location\=https%3A%2F%2Fcms.prx.org%2Fpub%2F572dbb04874a32676784039e7afc71eb%2F0%2Fweb%2Faudio_file%2F829012%2Fbroadcast%2FMoth1501Promo.mp3\&prx_story_id\=141178\&referrer\=https%3A%2F%2Fapi-sandbox.pmp.io%2Fdocs%2Fca7b2d27-f1e0-49ef-b831-b90a546cbfb5".

That seems to be b/c the audio shortcode implementation is relying on a check of the type based on the filename (basically seeing if the filename ends in `.mp3` or other audio extensions associated with  recognized/allowed mime types):
https://core.trac.wordpress.org/browser/tags/4.2.1/src/wp-includes/media.php#L1578
https://core.trac.wordpress.org/browser/tags/4.2.1/src/wp-includes/functions.php#L2049

So, it seems like the resolution here is that the player created for the embed or audio short code will work with the fix I am making to shorten the urls.

**Comment by cavis** [comment]
Okay, this is now working with current production audio for NPR, PRI and PRX.  Should eventually work just fine with APM.

Calling this one closed with 2 fallouts -> #61 #62
